actionpack/lib/action_dispatch/middleware/stack.rb

```diff
@@ -34,7 +34,28 @@ def inspect
       end
 
       def build(app)
-        klass.new(app, *args, &block)
+        InstrumentationProxy.new(klass.new(app, *args, &block), inspect)
+      end
+    end
+
+    # This class is used to instrument the execution of a single middleware.
+    # It proxies the `call` method transparently and instruments the method
+    # call.
+    class InstrumentationProxy
```

Comment:
Will these proxy objects show up in any user-facing API? Concerned whether we'll be obscuring the middleware stack anywhere. (If so, we could do something like generate proxy subclasses named after the middleware class.)

Reply:
We're already wrapping the middleware instances in Rails, I believe.